inc/class-job.php

```diff
@@ -219,6 +238,166 @@ public static function get_by_site( $site, $include_completed = false, $include_
+		}
+
+		if ( ! is_numeric( $args['site'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_site_id' );
+		}
+
+		if ( ! empty( $args['hook'] ) && ! is_string( $args['hook'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_hook_name' );
+		}
+
+		if ( ! is_array( $args['args'] ) && ! is_null( $args['args'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_event_arguments' );
+		}
+
+		if ( ! is_numeric( $args['limit' ] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_limit' );
+		}
+
+		if ( $args['timestamp'] === 'future' ) {
+			$timestamp = time();
+			$timestamp_compare = '>';
+		} elseif ( $args['timestamp'] === 'past' ) {
+			$timestamp = time();
+			$timestamp_compare = '<=';
+		}
+
+		$args['limit' ] = absint( $args['limit' ] );
+
+		if ( $args['limit'] > 100 ) {
+			trigger_error( 'Exceeding recommended job search limit of 100' );
+		}
+
+		// Find all scheduled events for this site
+		$table = static::get_table();
+
+		$sql = "SELECT * FROM `{$table}` WHERE site = %d";
+		$sql_params[] = $args['site'];
+
+		if ( is_string( $args['hook'] ) ) {
+			$sql .= ' AND hook = %s';
+			$sql_params[] = $args['hook'];
+		}
+
+		if ( ! is_null( $args['args'] ) ) {
+			$sql .= ' AND args = %s';
+			$sql_params[] = serialize( $args['args'] );
+		}
+
+		if (
+			! empty( $args['timestamp'] ) &&
+			in_array( $args['timestamp'], ['past', 'future'] ) &&
+			! empty( $timestamp ) &&
+			! empty( $timestamp_compare ) &&
+			in_array( $timestamp_compare, [ '<=', '<', '>', '>=', '=' ] )
+		) {
+			$sql .= " AND nextrun {$timestamp_compare} %s";
+			$sql_params[] = date( DATE_FORMAT, $timestamp );
+		} elseif (
+			! empty( $args['timestamp'] ) &&
+			is_array( $args['timestamp'] ) &&
+			count( $args['timestamp'] ) === 2
+		) {
+			// It's a range.
+			$sql .= ' AND nextrun BETWEEN %s AND %s';
+			$sql_params[] = date( DATE_FORMAT, $args['timestamp'][0] );
+			$sql_params[] = date( DATE_FORMAT, $args['timestamp'][1] );
+		} elseif ( ! empty( $args['timestamp'] ) ) {
+			$sql .= ' AND nextrun = %s';
+			$sql_params[] = date( DATE_FORMAT, $args['timestamp'] );
+		}
+
+		$sql .= ' AND status IN(' . implode( ',', array_fill( 0, count( $args['statuses'] ), '%s' ) ) . ')';
+		$sql_params = array_merge( $sql_params, $args['statuses'] );
+
+		$sql .= ' ORDER BY nextrun';
+		if ( $args['order'] === 'DESC' ) {
+			$sql .= ' DESC';
+		} else {
+			$sql .= ' ASC';
+		}
+
+		if ( $args['limit'] > 0 ) {
+			$sql .= ' LIMIT %d';
+			$sql_params[] = $args['limit'];
+		}
+
+		$query = $wpdb->prepare( $sql, $sql_params );
+
+		// Use caching if group deletion is supported.
+		if ( function_exists( 'wp_cache_delete_group' ) ) {
+			$query_hash = md5( $query );
+			$found = null;
+			$results = wp_cache_get( "jobs::{$query_hash}", 'cavalcade-job-queries', true, $found );
+			if ( ! $found ) {
+				$results = $wpdb->get_results( $query );
+				wp_cache_set( "jobs::{$query_hash}", $results, 'cavalcade-job-queries' );
+			}
```

Comment:
@peterwilsoncc can I get your input here? As we're using `wp-redis` in Altis we can use its cache group deletion feature.

Reply:
I think you can use the last changed pattern WP uses for similar queries to widen support. Something along the lines of the following untested code:

```php
// Cache result set.
if ( ! function_exists( 'wp_cache_delete_group' ) ) {
	$last_changed = '::' . wp_cache_get_last_changed( 'cavalcade-job-queries' );
}
$query_hash = md5( $query . $last_changed ?? '' );
$found = null;
$results = wp_cache_get( "jobs::{$query_hash}", 'cavalcade-job-queries', true, $found );

if ( ! $found ) {
	$results = $wpdb->get_results( $query );
	wp_cache_set( "jobs::{$query_hash}", $results, 'cavalcade-job-queries' );
}
```

With regard to the query hash: is `$wpdb->prepare()` expensive to run? If it is, setting the query hash to `md5( serialize( [ $sql, $sql_params ] ) )` might be quicker. Note: conjecture coming from a place of ignorance.